codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessageEncoder.java

```diff
@@ -0,0 +1,126 @@
+public final class HAProxyMessageEncoder extends MessageToByteEncoder<HAProxyMessage> {
+
+    private static final int V2_VERSION_BITMASK = 0x02 << 4;
+
+    // Length for source/destination addresses for the UNIX family must be 108 bytes each.
+    static final int UNIX_ADDRESS_BYTES_LENGTH = 108;
+    static final int TOTAL_UNIX_ADDRESS_BYTES_LENGTH = UNIX_ADDRESS_BYTES_LENGTH * 2;
+
+    public static final HAProxyMessageEncoder INSTANCE = new HAProxyMessageEncoder();
+
+    private HAProxyMessageEncoder() {
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, HAProxyMessage msg, ByteBuf out) throws Exception {
+        if (msg.protocolVersion() == HAProxyProtocolVersion.V1) {
+            encodeV1(msg, out);
+        } else if (msg.protocolVersion() == HAProxyProtocolVersion.V2) {
+            encodeV2(msg, out);
+        } else {
+            throw new HAProxyProtocolException("Unsupported version: " + msg.protocolVersion());
+        }
+    }
+
+    private static void encodeV1(HAProxyMessage msg, ByteBuf out) {
+        final String protocol = msg.proxiedProtocol().name();
+        StringBuilder sb = new StringBuilder(108)
+                .append("PROXY ").append(protocol).append(' ')
+                .append(msg.sourceAddress()).append(' ')
+                .append(msg.destinationAddress()).append(' ')
+                .append(msg.sourcePort()).append(' ').append(msg.destinationPort()).append("\r\n");
+        out.writeCharSequence(sb.toString(), CharsetUtil.US_ASCII);
+    }
+
+    private static void encodeV2(HAProxyMessage msg, ByteBuf out) {
+        out.writeBytes(BINARY_PREFIX);
+        out.writeByte(V2_VERSION_BITMASK | msg.command().byteValue());
+        out.writeByte(msg.proxiedProtocol().byteValue());
+
+        switch (msg.proxiedProtocol().addressFamily()) {
+            case AF_IPv4:
+            case AF_IPv6:
+                byte[] sourceAddress = NetUtil.createByteArrayFromIpAddressString(msg.sourceAddress());
+                byte[] destinationAddress = NetUtil.createByteArrayFromIpAddressString(msg.destinationAddress());
+                out.writeShort(sourceAddress.length + destinationAddress.length + 4 + msg.tlvNumBytes());
+                out.writeBytes(sourceAddress);
+                out.writeBytes(destinationAddress);
+                out.writeShort(msg.sourcePort());
+                out.writeShort(msg.destinationPort());
+                encodeTlvs(msg.tlvs(), out);
+                break;
+            case AF_UNIX:
+                out.writeShort(TOTAL_UNIX_ADDRESS_BYTES_LENGTH + msg.tlvNumBytes());
+                byte[] srcAddressBytes = msg.sourceAddress().getBytes(CharsetUtil.US_ASCII);
+                out.writeBytes(srcAddressBytes);
+                out.writeBytes(new byte[UNIX_ADDRESS_BYTES_LENGTH - srcAddressBytes.length]);
+                byte[] dstAddressBytes = msg.destinationAddress().getBytes(CharsetUtil.US_ASCII);
+                out.writeBytes(dstAddressBytes);
```

Comment:
why you can't just use `writeCharSequence(...)` here ? It seems there is no need to convert to `byte[]` first. Same comment is true for other places.

Reply:
I'm still looking for a more efficient way to do this.
The reason I can't just switch to `writeCharSequence` because I need to know the number of bytes to pad to length 108.
https://github.com/netty/netty/blob/b86a6c2e49c37b4872f075debddc06a9adcc4670/codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessageEncoder.java#L93

Any ideas are welcome

Reply:
`writeCharSequence` will return the number of bytes written. So You can use this, or I am missing something ?

Reply:
yea...   I don't read javadocs carefully 🤦